lib/src/internal/copy_on_write_set.dart

```diff
@@ -180,8 +165,27 @@ class CopyOnWriteSet<E> implements Set<E> {
   void _maybeCopyBeforeWrite() {
     if (!_copyBeforeWrite) return;
     _copyBeforeWrite = false;
-    _set = _setFactory != null
-        ? (_setFactory!()..addAll(_set))
-        : Set<E>.from(_set);
+    _set =
+        _setFactory != null ? (_setFactory!()..addAll(_set)) : Set<E>.of(_set);
+  }
+}
+
+/// Iterator wrapper on the [_wrapper._set.iterator].
+///
+/// Throws concurrent modification if a copy-on-write
+/// copy was made (assuming that a write was also made).
+class _CopyOnWriteSetIterator<E> implements Iterator<E> {
+  final Iterator<E> _source;
+  final CopyOnWriteSet<Object?> _wrapper;
+  _CopyOnWriteSetIterator(this._wrapper, this._source);
+  @override
+  bool moveNext() {
+    if (_wrapper._copyBeforeWrite) {
+      return _source.moveNext();
+    }
+    throw ConcurrentModificationError(_wrapper);
```

Comment:
This is probably too drastic. It considers it a concurrent modification even if the elements never changed (like doing `remove(notInSet)`).

That's avoidable by making the `CopyOnWriteSet` recognize which operations are no-ops first, and not copy until a real change is performed.
(For remove, do `if (!_set.contains(element)) return; _copy(); _set.remove(element);`, similarly or more complicated for all the other mutating operations.)